### PR TITLE
Add io.github.slgobinath.SafeEyes

### PR DESCRIPTION
**Pull request by otuva**
### Please confirm your submission meets all the criteria

- [X] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [X] My pull request follows the instructions at [App Submission][submission].
- [X] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [X] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [X] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub.
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [X] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

resolves slgobinath/SafeEyes#432

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

File: com.github.slgobinath.SafeEyes.metainfo.xml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop-application">
+    <id>com.github.slgobinath.SafeEyes</id>
```
Comment (hfiguiere) [inline com.github.slgobinath.SafeEyes.metainfo.xml:3]: it should `io.github` as per the documentation.

File: com.github.slgobinath.SafeEyes.yaml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+  - name: wlrctl # required for 'wayland do not disturb' plugin
+    buildsystem: meson
+    sources:
+      - type: archive
+        url: "https://git.sr.ht/~brocellous/wlrctl/archive/v0.2.1.tar.gz"
+        sha256: de8848de0a81749fe7ee7316bead0261bd978e26c2f899d02897609972a6e98c
+
+  - name: alsa-utils # required for 'audible alert' plugin
+    buildsystem: simple
```
Comment (hfiguiere) [inline com.github.slgobinath.SafeEyes.yaml:62]: this should be `autotools` buildsystem.

File: com.github.slgobinath.SafeEyes.yaml [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  - /lib/pkgconfig
+  - /share/gtk-doc
+  - /share/man
+  - /lib/*.a
+  - /lib/*.la
+
+modules:
+  - shared-modules/libappindicator/libappindicator-gtk3-introspection-12.10.json # app indicator support
+  - shared-modules/libappindicator/libappindicator-gtk3-12.10.json
+  - libcanberra/libcanberra.json # don't clean /lib/gtk-3.0
```
Comment (hfiguiere) [inline com.github.slgobinath.SafeEyes.yaml:39]: why do you not use the shared-module?

File: com.github.slgobinath.SafeEyes.yaml [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+command: safeeyes
+
+finish-args:
+  - "--share=network"
+  - "--share=ipc"
+  - "--socket=x11"
+  - "--socket=pulseaudio"
+  - "--device=dri"
+  - "--system-talk-name=org.freedesktop.DBus" # for tray icon
+  - "--talk-name=org.freedesktop.Notifications" # for 'notification before breaks' plugin
```
Comment (hfiguiere) [inline com.github.slgobinath.SafeEyes.yaml:14]: are you sure? libnotification 0.8.2 uses the portal and doesn't need any permission

**Review by hfiguiere — CHANGES_REQUESTED** [review]
have the desktop appstream and icon submitted upstream?

File: com.github.slgobinath.SafeEyes.metainfo.xml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    <name>Safe Eyes</name>
+    <developer_name>Gobinath</developer_name>
+    <summary>A Free and Open Source tool for Linux users to reduce and prevent repetitive strain
+        injury (RSI).</summary>
+
+    <metadata_license>CC-BY-4.0</metadata_license>
```
Comment (hfiguiere) [inline com.github.slgobinath.SafeEyes.metainfo.xml:10]: `CC0-1.0` is prefered.

File: com.github.slgobinath.SafeEyes.yaml (hunk 2 above)
Comment (otuva) [reply com.github.slgobinath.SafeEyes.yaml:62]: autotools without build-commands right?
Without prefix, build will fail 

```
/usr/bin/install: cannot create regular file '/lib/udev/rules.d/90-alsa-restore.rules': Read-only file system
```

File: com.github.slgobinath.SafeEyes.yaml [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+runtime-version: "22.08"
+command: safeeyes
+
+finish-args:
+  - "--share=network"
+  - "--share=ipc"
+  - "--socket=x11"
+  - "--socket=pulseaudio"
+  - "--device=dri"
+  - "--system-talk-name=org.freedesktop.DBus" # for tray icon
```
Comment (hfiguiere) [inline com.github.slgobinath.SafeEyes.yaml:13]: I don't think this VERY BROAD permission is neceary.

File: com.github.slgobinath.SafeEyes.yaml (hunk 3 above)
Comment (otuva) [reply com.github.slgobinath.SafeEyes.yaml:39]: After this [commit](https://github.com/flathub/shared-modules/commit/096ab6b0be271f07a42a647b5ffc83808a160e59), installation deletes `/lib/gtk-3.0` too so instead of forking it I'm just using the same without the gtk-3.0 line

Without  `/lib/gtk-3.0`, app will give a warning `Failed to load module "canberra-gtk-module"` probably it's not that important but I didn't check the whole codebase and upstream says it's a requirement

File: com.github.slgobinath.SafeEyes.yaml (hunk 4 above)
Comment (otuva) [reply com.github.slgobinath.SafeEyes.yaml:14]: Tried a couple of times and it didn't work without it. Let me try again though

**Comment by otuva** [comment]
> have the desktop appstream and icon submitted upstream?

Icon is from the upstream, but I don't see why the XML file should be submitted. I can though if it's necessary

**Comment by hfiguiere** [comment]
> > have the desktop appstream and icon submitted upstream?
> 
> Icon is from the upstream, but I don't see why the XML file should be submitted. I can though if it's necessary

the the icon should NOT be in this repository if it is upstream. How about the .desktop?

Yes the appstream file too should be submitted upstream. It is not flatpak specific either.

File: com.github.slgobinath.SafeEyes.yaml (hunk 3 above)
Comment (hfiguiere) [reply com.github.slgobinath.SafeEyes.yaml:39]: this is how all the packages that use it use it.

**Comment by hfiguiere** [comment]
the files need to be renamed too.

File: com.github.slgobinath.SafeEyes.yaml (hunk 2 above)
Comment (hfiguiere) [reply com.github.slgobinath.SafeEyes.yaml:62]: `config-opts`

**Comment by otuva** [comment]
> the files need to be renamed too.

Should I open new PR with io one to resolve discrepancy between file reverse dns and repo? I saw a bunch of packages following com.github so that's why I did that

**Comment by hfiguiere** [comment]
just rename the files.

**Comment by hfiguiere** [comment]
>  I saw a bunch of packages following com.github so that's why I did that

This rule was not enforced before. But it's in the documentation.

File: com.github.slgobinath.SafeEyes.yaml (hunk 6 above)
Comment (otuva) [reply com.github.slgobinath.SafeEyes.yaml:13]: Is it ok to use `org.freedesktop.StatusNotifierItem.Menu` ?

File: com.github.slgobinath.SafeEyes.yaml [hunk 7]
```diff
@@ -0,0 +1,99 @@
+id: com.github.slgobinath.SafeEyes
```
Comment (hfiguiere) [inline com.github.slgobinath.SafeEyes.yaml:1]: this also doesn't match. The app id has to be changed everywhere.

Comment (otuva) [reply com.github.slgobinath.SafeEyes.yaml:1]: Yeah, I'm going to push the commit that changes them. But currently I couldn't figure out one thing. 

So, the app, after setup, installs icons and desktop file under `/app/lib/python3.10/site-packages/safeeyes/platform/` I couldn't find any docs or different packages related to this. How do you think I should proceed? `install`, mv, cp or symlinks don't feel 'clean'

File: io.github.slgobinath.SafeEyes.metainfo.xml [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        </p>
+        <p>
+            Remind you to take breaks with exercises to reduce RSI, Disable keyboard during breaks,
+            Notification before and after breaks, Smart pause if system is idle, Multi-screen
+            support, Customizable user interface, RPC API to control externally, Command-line
+            arguments to control the running instance, Customizable using plug-ins
+        </p>
+    </description>
+
+    <launchable type="desktop-id">com.github.slgobinath.SafeEyes.desktop</launchable>
```
Comment (hfiguiere) [inline io.github.slgobinath.SafeEyes.metainfo.xml:34]: this doesn't match.

File: io.github.slgobinath.SafeEyes.yaml [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+          project-id: 13149
+          type: anitya
+          url-template: https://download.gnome.org/sources/libnotify/$major.$minor/libnotify-$version.tar.xz
+
+  - name: safeeyes # shortcut
+    buildsystem: simple
+    build-commands:
+      - install -Dm644 /app/lib/python3.10/site-packages/safeeyes/platform/icons/hicolor/128x128/apps/safeeyes.png /app/share/icons/hicolor/128x128/apps/io.github.slgobinath.SafeEyes.png
+      - install -Dm644 /app/lib/python3.10/site-packages/safeeyes/platform/safeeyes.desktop /app/share/applications/io.github.slgobinath.SafeEyes.desktop
+      - desktop-file-edit --rebuild-mime-info-cache --set-icon=io.github.slgobinath.SafeEyes --set-key=Exec --set-value=safeeyes /app/share/applications/io.github.slgobinath.SafeEyes.desktop
```
Comment (hfiguiere) [inline io.github.slgobinath.SafeEyes.yaml:91]:
```suggestion
      - desktop-file-edit --set-icon=io.github.slgobinath.SafeEyes --set-key=Exec --set-value=safeeyes /app/share/applications/io.github.slgobinath.SafeEyes.desktop
```

File: io.github.slgobinath.SafeEyes.yaml [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+  - pypi-dependencies.json
+
+  - name: dbus-python # dbus - for tray
+    buildsystem: autotools
+    sources:
+      - type: archive
+        url: "https://dbus.freedesktop.org/releases/dbus-python/dbus-python-1.3.2.tar.gz"
+        sha256: ad67819308618b5069537be237f8e68ca1c7fcc95ee4a121fe6845b1418248f8
+
+  - name: wlrctl # required for 'wayland do not disturb' plugin
```
Comment (hfiguiere) [inline io.github.slgobinath.SafeEyes.yaml:53]: why do you need this if you don't have the wayland socket.

File: io.github.slgobinath.SafeEyes.yaml [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    buildsystem: meson
+    sources:
+      - type: archive
+        url: "https://git.sr.ht/~brocellous/wlrctl/archive/v0.2.1.tar.gz"
+        sha256: de8848de0a81749fe7ee7316bead0261bd978e26c2f899d02897609972a6e98c
+
+  - name: alsa-utils # required for 'audible alert' plugin
+    buildsystem: autotools
+    config-opts:
+      - "--prefix=/app"
```
Comment (hfiguiere) [inline io.github.slgobinath.SafeEyes.yaml:63]:
```suggestion
```

File: io.github.slgobinath.SafeEyes.yaml (hunk 10 above)
Comment (otuva) [reply io.github.slgobinath.SafeEyes.yaml:53]: Even though the app only runs through x11, it can still detect wayland and `wlrctl` is required to detect full screen

Comment (hfiguiere) [reply io.github.slgobinath.SafeEyes.yaml:53]: how does it detect wayland?

Comment (otuva) [reply io.github.slgobinath.SafeEyes.yaml:53]: with [this function](https://github.com/slgobinath/SafeEyes/blob/cae19b0f28f14036d1a2a4c840bf3e6f68c02b6f/safeeyes/utility.py#L299) 

without wlrctl you can see [the warning](https://github.com/slgobinath/SafeEyes/blob/cae19b0f28f14036d1a2a4c840bf3e6f68c02b6f/safeeyes/plugins/donotdisturb/plugin.py#L52)

Comment (hfiguiere) [reply io.github.slgobinath.SafeEyes.yaml:52]: do you test is inside the flatpak sandbox? Because this detection is broken.

**Comment by hfiguiere** [comment]
bot, build io.github.slgobinath.SafeEyes

File: io.github.slgobinath.SafeEyes.yaml (hunk 10 above)
Comment (otuva) [reply io.github.slgobinath.SafeEyes.yaml:52]: Yeah, you're right. I'm going to keep to fiddle with it. It probably acts like that because of missing wayland socket

File: flathub.json [hunk 12]
```diff
@@ -0,0 +1,5 @@
+{
+    "only-arches": [
+        "x86_64"
```
Comment (hfiguiere) [inline flathub.json:3]: you could just fix the broken code instead.

Comment (otuva) [reply flathub.json:3]: Code should be 

```
bool is_ascii(const char str[])
{
    for (int i = 0; str[i] != '\0'; i++) {
        if (str[i] > 127) {
            return false;
        }
    }
    return true;
}
```

right? How do I patch that though, can you supply patch even if it's meson

Comment (otuva) [reply flathub.json:3]: Apparently author of the package marked it as x86_64 arch only

[aur pkgbuild](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=wlrctl)

Comment (hfiguiere) [reply flathub.json:3]: that will not build on x86_64.

Comment (hfiguiere) [reply flathub.json:3]: > Apparently author of the package marked it as x86_64 arch only
> 
> [aur pkgbuild](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=wlrctl)

then don't build this on aarch64.

Comment (hfiguiere) [reply flathub.json:3]: and by this I mean wlrctl.

**Comment by otuva** [comment]
Hi, @hfiguiere, sorry to bother you mate, but what is the mergeability status of this?

**Comment by otuva** [comment]
> Shouldn't it create some default config on start?

Yeah, it should, it stems from [this part](https://github.com/slgobinath/SafeEyes/blob/master/safeeyes/utility.py#L377) and author is aware of the problem. I initially wanted to fix it in manifest, but evidently in docs there's no way to create config files for the user.

I added to the README for the app's repo, but probably no one is going to read that.  

I'm going to submit a patch to the upstream, until then if you think it's a big problem I can write a wrapper for initial launch.
